app/models/rbac.rb

```diff
@@ -331,68 +329,48 @@ def self.filtered(objects, options = {})
     Rbac.search(options.merge(:targets => objects, :results_format => :objects, :empty_means_empty => true)).first
   end
 
+  # @param klass [Class] base_class found in CLASSES_THAT_PARTICIPATE_IN_RBAC
   def self.find_via_descendants(descendants, method_name, klass)
-    matches = []
-    descendants.each do |object|
+    MiqPreloader.preload(descendants, method_name)
+    descendants.flat_map do |object|
       match = object.send(method_name)
       match = [match] unless match.kind_of?(Array)
-      match.each do |m|
-        next unless m.kind_of?(klass)
-        next if matches.include?(m)
-        matches << m
-      end
-    end
-    matches
-  end
-
-  def self.find_descendants(descendant_klass, options)
-    search(options.merge(:class => descendant_klass, :results_format => :objects)).first
+      match.select { |m| m.kind_of?(klass) }
+    end.uniq
```

Comment:
yea, thanks - I want to move that `grep`/`select` into the original query, so others are not even brought back.